### DISCUSSION TITLE
range support on nameof for compiled time adjusted strings

### DISCUSSION THREAD

**Discussion by msschl** [dotnet/csharplang#5305] [Language Ideas]
It would be nice if the `nameof(T)` operator would support some simple range syntax to store a modified string at compile time.

A simple use-case would be for creating a string from a type e.g. `TestEndpoint` and while displaying the endpoint name the `Endpoint` suffix should be omitted.
I have setup a simple sample in [sharplab.io](https://sharplab.io/#v2:D4AQTAjAsAULIGYAE4kGEkG9ZNy5IALEgLIAUAlFjnrQG4CGATkgEYCuAlgDYAmApiwC8SAHb8A7kgAiASQBCXPoMoBuGrVwbNICAE4yAIgiGK6mJrwceApgDoAgr14BRUbwAOAe06iALgA8aAB8Rh4MfgAWpuaWWhZxugaGYDHatNbK9k6u7t6+gWhunj7+oaIMALb8XgBmZEV5pX4UANp2dgB6ABwAugA0SIbhUWkJuAC+sFNwMIgoYOjF+f7U40jTsPAEi3KKNoKw2OvzREg5y80BACqhugAMYlX8gw9II5EU6XhCwSj6ZAAJIZMBVqhMkL8sB8JmMZtt/gA2BZIa78ADOfiO2lOEGReyUtnOzkuBRuoSinHRMgUhMEbDpTFeEEeHy+62OcSQjBYYP4kKQfgAnh4avVrhQ7AA5Z7tLp9WJxTK2Rwkppk25kPmDNmKyzfXAgADsDIOTD1SBmEyAA==)

```cs
using System;

public class C {
    public void M() {
        var builder = new DIBuilder();
        
        Console.WriteLine("1");
        builder.AddEndpoint<C>("path");
        
        Console.WriteLine("2");
        builder.AddEndpoint<CEndpoint>(nameof(CEndpoint)[..^8], "path");
    }
}

public class CEndpoint {
    
}

public class DIBuilder
{
    public void AddEndpoint<T>(string name, string path)
        => Console.WriteLine($"{name} => {path}");
}

public static class Test
{
    public static DIBuilder AddEndpoint<T>(this DIBuilder builder, string path)
    {
        var name = typeof(T).Name[..^8];
        builder.AddEndpoint<T>(name, path);
        
        return builder;
    }
}
```

# Proposal

Add a `nameof(Type, System.Range range)` overload and store the compile-time generated string accordingly.

```cs
var name = nameof(TestEndpoint, ..^8);
```

**Comment by CyrusNajmabadi** [comment]
1. why does this need to happen at compile time?
2. There is no guarantee that the range operation will work with this on any particular runtime.  For example, it might do something related to runtime locale that might affect the results.

I think the only way this would happen is if [Compile Time Expressions](https://github.com/dotnet/csharplang/discussions/1028) becomes a thing.  I don't think there would be anything special about this.
